File: Medium/0565-Array Nesting/Solution.rb
# @param {Integer[]} nums
# @return {Integer}
def array_nesting(nums)
  ret = 1

  (0...nums.size).each do |i|
    next if nums[i] < 0

    length = 0
    j = i
    while nums[j] >= 0
      nums[j] = -nums[j] - 1
      length += 1
      j = -nums[j] - 1
    end

    ret = [ret, length].max
  end

  ret
end

File: Medium/0565-Array Nesting/Solution.rs
impl Solution {
    pub fn array_nesting(mut nums: Vec<i32>) -> i32 {
        let mut ret = 1;

        for i in 0..nums.len() {
            if nums[i] < 0 {
                continue;
            }

            let mut length = 0;
            let mut j = i;
            while nums[j] >= 0 {
                nums[j] = -nums[j] - 1;
                length += 1;
                j = (-nums[j] - 1) as usize;
            }

            ret = ret.max(length);
        }

        ret
    }
}
